Support zero dimensions in softmax decompositions

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #91322

The eager implementation of softmax supports computation along zero dimensions, but many of the other implementations did not, including:
* decompositions & refs (this was causing dynamo failures)
* forward AD for logsumexp
* MPS log_softmax_backward

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions, and updates opinfos.

example of "computation along zero dimensions":

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

davidberard98 committed:
[WIP] fix softmax decomposition handling of zero dimensions

```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] fix softmax decomposition handling of zero dimensions"


```python
import torch
import torch.nn.functional as F

t = torch.rand((4, 0, 0))
print("~")
print(F.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=0)  # this fails
print("~")
```

[ghstack-poisoned]

davidberard98 committed:
Update on "Support zero dimensions in softmax decompositions"


The eager implementation of softmax supports computation along zero dimensions, but decompositions & forward AD implementations did not, like shown below.

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions.

[ghstack-poisoned]

davidberard98 commented:
cc @lezcano who I think has been doing related work recently?

davidberard98 commented:
nvm, looks like the mps failure is legitimate; going to take a look at that first

davidberard98 committed:
Update on "Support zero dimensions in softmax decompositions"


The eager implementation of softmax supports computation along zero dimensions, but decompositions & forward AD implementations did not, like shown below.

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions.

[ghstack-poisoned]

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -987,6 +987,8 @@ def _softmax(x: Tensor, dim: int, half_to_float: bool):
     computation_dtype, result_dtype = utils.elementwise_dtypes(
         x, type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
+    if x.numel() == 0:
+        return torch.zeros_like(x, dtype=result_dtype)
```
See that the cast to `result_type` just happens when `not half_to_float`.

To avoid these issues, I would suggest to simply set `x_max = zeros_like(x)` in this case, rather than trying to completely short cut the computation in this case.

davidberard98 replied:
thanks! let me update that.

lezcano reviewed (commented):
A small point.

davidberard98 committed:
Update on "Support zero dimensions in softmax decompositions"


The eager implementation of softmax supports computation along zero dimensions, but decompositions & forward AD implementations did not, like shown below.

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions.

[ghstack-poisoned]

davidberard98 committed:
Update on "Support zero dimensions in softmax decompositions"


The eager implementation of softmax supports computation along zero dimensions, but decompositions & forward AD implementations did not, like shown below.

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions.

[ghstack-poisoned]

davidberard98 committed:
Update on "Support zero dimensions in softmax decompositions"


The eager implementation of softmax supports computation along zero dimensions, but many of the other implementations did not, including:
* decompositions & refs (this was causing dynamo failures)
* forward AD for logsumexp
* MPS log_softmax_backward

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions, and updates opinfos.

example of "computation along zero dimensions":

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

[ghstack-poisoned]

lezcano reviewed (approved):
LGTM!

lezcano commented:
@pytorchbot merge -f "unrelated error"
